### PR TITLE
Feat node cli

### PR DESCRIPTION
**Pull request by christyjacob4**

### CODE REVIEW

**Comment by lohanidamodar** [comment]
@christyjacob4 please sync with `feat-preps-for-0.13` and also switch base to `feat-preps-for-0.13` as the new CLI is using features from 0.13 (large file upload)

File: templates/node-cli/lib/parser.js.twig [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+const logo = "\n" +
+    "    _                            _ _           ___   __   _____ \n" +
+    "   /_\\  _ __  _ ____      ___ __(_) |_ ___    / __\\ / /   \\_   \\\n" +
+    "  //_\\\\| '_ \\| '_ \\ \\ /\\ / / '__| | __/ _ \\  / /   / /     / /\\/\n" +
+    " /  _  \\ |_) | |_) \\ V  V /| |  | | ||  __/ / /___/ /___/\\/ /_  \n" +
+    " \\_/ \\_/ .__/| .__/ \\_/\\_/ |_|  |_|\\__\\___| \\____/\\____/\\____/  \n" +
+    "       |_|   |_|                                                  \n" +
+    "\n";
```
Comment (eldadfux) [inline templates/node-cli/lib/parser.js.twig:155]: Again, should be a twig variable.

File: templates/node-cli/lib/parser.js.twig [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+const commandDescriptions = {
+    "account": `The account command allows you to authenticate and manage a user account.`,
+    "avatars": `The avatars command aims to help you complete everyday tasks related to your app image, icons, and avatars.`,
+    "database": `The database command allows you to create structured collections of documents, query and filter lists of documents.`,
+    "deploy": `The deploy command provides a convenient wrapper for deploying your functions and collections.`,
+    "functions": `The functions command allows you view, create and manage your Cloud Functions.`,
+    "health": `The health command allows you to both validate and monitor your {{ spec.title|caseUcfirst }} server's health.`,
+    "init": `The init command helps you initialize your {{ spec.title|caseUcfirst }} project, functions and collections`,
+    "locale": `The locale command allows you to customize your app based on your users' location.`,
```
Comment (eldadfux) [inline templates/node-cli/lib/parser.js.twig:165]: Can we generate some of this block?

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/SDK/SDK.php [hunk 3]
```diff
@@ -58,6 +58,7 @@ class SDK
         'readme' => '',
         'changelog' => '',
         'examples' => '',
+        'isTest' => 'false'
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/SDK.php:61]: We use is to selectively remove parts of the template during tests. 
Client functions like `setKey` and `setMode` are not part of the mock spec and leads to errors during tests.

[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/SDK.php]: done 👍

File: src/SDK/Language/NodeCLI.php [hunk 4]
```diff
@@ -0,0 +1,223 @@
+<?php
+
+namespace Appwrite\SDK\Language;
+
+class NodeCLI extends JS
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: Done 👍  Removed old CLI and renamed this to CLI

[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: done 👍

File: src/SDK/Language/NodeCLI.php [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    /**
+     * @return array
+     */
+    public function getFiles()
+    {
+        return [
+            [
+                'scope'         => 'default',
+                'destination'   => 'README.md',
+                'template'      => 'node-cli/README.md.twig',
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: done 👍

File: src/SDK/Language/NodeCLI.php [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+        return $this;
+    }
+
+    /**
+     * @return string
+     */
+    public function getName()
+    {
+        return 'node-cli';
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: done 👍

File: src/SDK/Language/NodeCLI.php [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            ],
+            [
+                'scope'         => 'default',
+                'destination'   => '.gitignore',
+                'template'      => 'node-cli/.gitignore',
+                'minify'        => false,
+            ],
+            [
+                'scope'         => 'copy',
+                'destination'   => 'Formula/appwrite.rb',
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: @wess Can clarify. Im not completely sure about this

File: src/SDK/Language/NodeCLI.php [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+                'destination'   => 'lib/commands/init.js',
+                'template'      => 'node-cli/lib/commands/init.js.twig',
+                'minify'        => false,
+            ],
+            [
+                'scope'         => 'default',
+                'destination'   => 'lib/commands/deploy.js',
+                'template'      => 'node-cli/lib/commands/deploy.js.twig',
+                'minify'        => false,
+            ],
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: Understood 👍

File: src/SDK/Language/NodeCLI.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (wess) [reply src/SDK/Language/NodeCLI.php]: We need appwrite to be the forumla name for Homebrew. Homebrew uses it for install, search and even upgrading/updating.

File: templates/node-cli/lib/parser.js.twig (hunk 2 above)
Comment (christyjacob4) [reply templates/node-cli/lib/parser.js.twig]: Unfortunately the description of the services is not part of the swagger spec. Unless we add it to the spec itself, it doesn't seem like we can generate this .

File: templates/node-cli/install.sh.twig [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+# (and let us know you found this message...)
+
+# This script contains hidden JS code to allow better readability and syntax highlighting
+# You can use "View source" of this page to see the full script.
+
+# {{ spec.title | caseUcfirst }} CLI location
+{{ spec.title | upper }}_INSTALL_DIR="/usr/local/bin"
+
+# {{ spec.title | caseUcfirst }} CLI Executable name 
+{{ spec.title | upper }}_EXECUTABLE_NAME=appwrite
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply templates/node-cli/install.sh.twig]: replaced 👍

File: templates/node-cli/install.sh.twig [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    printf "${GREEN}🚦 Fetching latest version ... ${NC}\n"
+    res=$(curl -L -s -H 'Accept: application/json' https://github.com/$GITHUB_REPOSITORY_NAME/releases/latest)
+    if [[ "$res" == *"error"* ]]; then
+        printf "${RED}❌ There was an error. Try again later.${NC} \n"
+        exit 1
+    fi
+    printSuccess
+
+    GITHUB_LATEST_VERSION=$( echo $res | sed -e 's/.*"tag_name":"\([^"]*\)".*/\1/')
+    GITHUB_FILE="appwrite-cli-${OS}-${ARCH}"
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply templates/node-cli/install.sh.twig]: replaced 👍

File: templates/node-cli/Formula/appwrite.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+require "language/node"
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply templates/node-cli/Formula/appwrite.rb]: fixed 👍

File: templates/node-cli/README.md.twig [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+If you do not have `npm` installed, you can always install the prebuilt binaries for your architecture and OS using our convenient installation scripts.
+
+### Linux / MacOS Terminal
+```bash
+$ wget -q {{ sdk.url }}/cli/install.sh  -O - | /bin/bash
+```
+
+### MacOS via [Homebrew](https://brew.sh)
+```bash
+$ brew tap appwrite/sdk-for-node-cli https://github.com/appwrite/sdk-for-node-cli
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply templates/node-cli/README.md.twig]: we can remove it once we have 30+ stars, forks and a community member raises a PR to the brew  central repository

File: templates/node-cli/README.md.twig [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+3. Head over to the generated SDK and install the dependencies.
+```sh
+$ cd examples/node-cli
+$ npm install 
+```
+
+4. Install the CLI using 
+```sh
+$ npm install -g .
+```
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply templates/node-cli/README.md.twig]: We could remove it but this was added to the cli readme after feedback from the community that they needed instructions for local setup and contributing within the repo

File: templates/node-cli/lib/parser.js.twig (hunk 1 above)
Comment (christyjacob4) [reply templates/node-cli/lib/parser.js.twig]: Done

File: templates/node-cli/install.sh.twig [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+greeting() {
+    echo -e "${RED}"
+    cat << "EOF"
+    _                            _ _           ___   __   _____ 
+   /_\  _ __  _ ____      ___ __(_) |_ ___    / __\ / /   \_   \
+  //_\\| '_ \| '_ \ \ /\ / / '__| | __/ _ \  / /   / /     / /\/
+ /  _  \ |_) | |_) \ V  V /| |  | | ||  __/ / /___/ /___/\/ /_  
+ \_/ \_/ .__/| .__/ \_/\_/ |_|  |_|\__\___| \____/\____/\____/  
+       |_|   |_|                                                  
+
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply templates/node-cli/install.sh.twig]: done

File: templates/node-cli/lib/parser.js.twig (hunk 2 above)
Comment (eldadfux) [reply templates/node-cli/lib/parser.js.twig]: I was sure we did it. See the `tags` attribute here: https://raw.githubusercontent.com/appwrite/appwrite/master/app/config/specs/swagger2-0.12.x-server.json

File: src/SDK/Language/NodeCLI.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (eldadfux) [reply src/SDK/Language/NodeCLI.php]: Got it. @christyjacob4 this is how we can give a file a dynamic name: https://github.com/appwrite/sdk-generator/blob/fed6a5a2276aeabbafce30565e4a51e0ec2f214f/src/SDK/Language/Android.php#L31

File: templates/node-cli/lib/parser.js.twig (hunk 2 above)
Comment (christyjacob4) [reply templates/node-cli/lib/parser.js.twig]: Screenshot 2022-02-24 at 2 05 00 PM
I believe tags give information about an endpoint and not about the entire service.

Comment (christyjacob4) [reply templates/node-cli/lib/parser.js.twig]: These descriptions are for the entire service as a whole

File: src/SDK/Language/NodeCLI.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (christyjacob4) [reply src/SDK/Language/NodeCLI.php]: updated 👍

**Pull request merged**